EPOParser.py

```diff
@@ -4,115 +4,133 @@
+def from_element(info):
+    """This is a docstring, which is used to comment on a function."""
+    return lambda el: getattr(el, info)
 
 # tree :: String -> Tree
 def tree(filename):
-  return etree.parse(filename, etree.XMLParser(encoding='utf-8', recover=True))
-
-# getFileInfo :: String -> [(filename, file-location)]
-def getFileInfo(indexFilename):
-  return fzip(
-    fmap(fromElement('text'), getElement(tree(indexFilename), '//filename')),
-    fmap(fromElement('text'), getElement(tree(indexFilename), '//file-location'))
-  )
-
-# getElement :: Tree, xpath -> [Element]
-def getElement(tree, elementPath):
-  return tree.xpath(elementPath)
-
-# getMetadata :: Tree, xpath, attrib | text, String -> [] | [(alias, tag, {} | String)]
-def getMetadata(tree, elementPath, info, alias):
-  return fmap(lambda el: ( alias, el.tag, getattr(el, info) ), getElement(tree, elementPath))
+    """
+    This is a multi-line docstring.
+
+    It is used when there is more info that should be shared than fits on one
+    line.
+    """
+    return etree.parse(
+        filename, etree.XMLParser(encoding='utf-8', recover=True))
+
+# get_file_info :: String -> [(filename, file-location)]
+def get_file_info(indexFilename):
+    return fzip(
+        fmap(from_element('text'), get_element(tree(indexFilename), '//filename')),
+        fmap(from_element('text'), get_element(tree(indexFilename), '//file-location'))
+    )
+
+# get_element :: Tree, xpath -> [Element]
+def get_element(tree, elementPath):
+    return tree.xpath(elementPath)
+
+# get_metadata :: Tree, xpath, attrib | text, String -> [] | [(alias, tag, {} | String)]
+def get_metadata(tree, elementPath, info, alias):
+    return fmap(lambda el: ( alias, el.tag, getattr(el, info) ), get_element(tree, elementPath))
 
 # parse :: Tree, [[ xpath, attrib | text, String ]] -> [(alias, tag {} | String)]
 def parse(tree, fields):
-  return fcompose(
-    fcmap(lambda x: x[0]),
-    fcfilter(lambda x: bool(x)),
-    fcmap(lambda field: getMetadata(tree, field[0], field[1], field[2]))
-  )(fields)
+    return fcompose(
+        fcmap(lambda x: x[0]),
+        fcfilter(lambda x: bool(x)),
+        fcmap(lambda field: get_metadata(tree, field[0], field[1], field[2]))
+    )(fields)
 
 ##############
 # processing #
 ##############
 
 def process(fields, state, callback, xmlFile, archiveFile):
-  if not os.path.isfile(marker(xmlFile)):
-    zipfile.ZipFile(archiveFile, "r").extract(xmlFile)
-    
+    if not os.path.isfile(marker(xmlFile)):
+        zipfile.ZipFile(archiveFile, "r").extract(xmlFile)
+
     # add validation/testing
-    
-    metadata = parse( tree(xmlFile), fields() )
-    
+
+    metadata = parse(tree(xmlFile), fields())
+
     # send off to queue (handle failures and restarts), if success, write file...
     callback(metadata)
-    
+
     # gives us a durable state if the script breaks and reruns
-    completefile = open(marker(xmlFile),"w") 
+    completefile = open(marker(xmlFile), "w")
```

Comment:
Here you're just writing a file one-off, so I think this is okay. However, `open` returns a stream that developers often forget to close. Thus, there is a block syntax that can be used to close out the file automatically. Often this is useful if you are opening the file so that it can be appended or writing to a new file over many function calls.

```
# This should append the date to completefile and automatically close the stream
with open(marker(xmlFile), "a") as completefile:
    completefile.write(datetime.datetime.now().strftime("%Y-%m-%d %H:%M:%S"))
```